{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Note\n",
    "This notebook needs to be tidied. There is a lot of mess and the purpose is unclear for some parts. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The main aspect is the reintroduction of the core SRA columns to the table that is outputted as clean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import sqlite3 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_117985/2623008537.py:5: DtypeWarning: Columns (37,39,40,41,63,72,73,78,81,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  filtered = pd.read_csv('/home/jack/projects/RiboSeqOrg-DataPortal/data/filtered_riboseq_done_260623.csv')\n"
     ]
    }
   ],
   "source": [
    "files = pd.read_csv('/home/jack/projects/RiboSeqOrg-DataPortal/data/trips_files.csv',\n",
    "                    names=['file_id', 'organism_id', 'study_id', 'file_name', 'file_description', 'file_type', 'owner', 'mapped_reads', 'control', 'cell_line'])\n",
    "studies = pd.read_csv('/home/jack/projects/RiboSeqOrg-DataPortal/data/trips_studies.csv',\n",
    "                      names=['study_id', 'organism_id', 'study_name', 'paper_authors', 'srp_nos', 'paper_year', 'paper_pmid', 'paper_link', 'gse_nos', 'adapters', 'paper_title', 'description', 'private', 'owner'])\n",
    "filtered = pd.read_csv('/home/jack/projects/RiboSeqOrg-DataPortal/data/filtered_riboseq_done_260623.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12725, 10)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>study_name</th>\n",
       "      <th>paper_authors</th>\n",
       "      <th>srp_nos</th>\n",
       "      <th>paper_year</th>\n",
       "      <th>paper_pmid</th>\n",
       "      <th>paper_link</th>\n",
       "      <th>gse_nos</th>\n",
       "      <th>adapters</th>\n",
       "      <th>paper_title</th>\n",
       "      <th>description</th>\n",
       "      <th>private</th>\n",
       "      <th>owner</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.0</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1.0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>Nedialkova15</td>\n",
       "      <td>Nedialkova DD, Leidel SA</td>\n",
       "      <td>SRP056647</td>\n",
       "      <td>2015</td>\n",
       "      <td>26052047</td>\n",
       "      <td>\"https://linkinghub.elsevier.com/retrieve/pii/...</td>\n",
       "      <td>GSE67387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Optimization of Codon Translation Rates via tR...</td>\n",
       "      <td>Ribosome profiling of wild-type and tRNA modif...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2.0</th>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>Dunn13</td>\n",
       "      <td>Dunn JG, Foo CK, Belletier NG, Gavis ER, Weiss...</td>\n",
       "      <td>SRP028243</td>\n",
       "      <td>2013</td>\n",
       "      <td>24302569</td>\n",
       "      <td>\"https://doi.org/10.7554/eLife.01179\"</td>\n",
       "      <td>GSE49197</td>\n",
       "      <td>CTGTAGGCACCATCAAT</td>\n",
       "      <td>Ribosome profiling reveals pervasive and regul...</td>\n",
       "      <td>12 samples of Drosophila ribosome profiling an...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.0</th>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>Nakahigashi16</td>\n",
       "      <td>Nakahigashi K, Takai Y, Kimura M, Abe N, Nakay...</td>\n",
       "      <td>DRP003075</td>\n",
       "      <td>2016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://academic.oup.com/dnaresearch/article-l...</td>\n",
       "      <td>Nakahigashi16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Comprehensive identification of translation st...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     study_id  organism_id     study_name   \n",
       "NaN         0            1              2  \\\n",
       "0.0        -1            0            NaN   \n",
       "1.0         1           10   Nedialkova15   \n",
       "2.0         2            7         Dunn13   \n",
       "3.0         3            8  Nakahigashi16   \n",
       "\n",
       "                                         paper_authors    srp_nos paper_year   \n",
       "NaN                                                  3          4          5  \\\n",
       "0.0                                                NaN        NaN        NaN   \n",
       "1.0                           Nedialkova DD, Leidel SA  SRP056647       2015   \n",
       "2.0  Dunn JG, Foo CK, Belletier NG, Gavis ER, Weiss...  SRP028243       2013   \n",
       "3.0  Nakahigashi K, Takai Y, Kimura M, Abe N, Nakay...  DRP003075       2016   \n",
       "\n",
       "    paper_pmid                                         paper_link   \n",
       "NaN          6                                                  7  \\\n",
       "0.0        NaN                                                NaN   \n",
       "1.0   26052047  \"https://linkinghub.elsevier.com/retrieve/pii/...   \n",
       "2.0   24302569              \"https://doi.org/10.7554/eLife.01179\"   \n",
       "3.0        NaN  https://academic.oup.com/dnaresearch/article-l...   \n",
       "\n",
       "           gse_nos           adapters   \n",
       "NaN              8                  9  \\\n",
       "0.0            NaN                NaN   \n",
       "1.0       GSE67387                NaN   \n",
       "2.0       GSE49197  CTGTAGGCACCATCAAT   \n",
       "3.0  Nakahigashi16                NaN   \n",
       "\n",
       "                                           paper_title   \n",
       "NaN                                                 10  \\\n",
       "0.0                                                NaN   \n",
       "1.0  Optimization of Codon Translation Rates via tR...   \n",
       "2.0  Ribosome profiling reveals pervasive and regul...   \n",
       "3.0  Comprehensive identification of translation st...   \n",
       "\n",
       "                                           description  private  owner  \n",
       "NaN                                                 11       12     13  \n",
       "0.0                                                NaN        0      0  \n",
       "1.0  Ribosome profiling of wild-type and tRNA modif...        0      1  \n",
       "2.0  12 samples of Drosophila ribosome profiling an...        0      1  \n",
       "3.0                                                NaN        0      1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "studies.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "public_studies = studies[studies['private'] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "public_study_ids = list(public_studies['study_id'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>spots</th>\n",
       "      <th>bases</th>\n",
       "      <th>avgLength</th>\n",
       "      <th>size_MB</th>\n",
       "      <th>LibraryName</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>...</th>\n",
       "      <th>Host</th>\n",
       "      <th>UMI</th>\n",
       "      <th>Adapter</th>\n",
       "      <th>Separation</th>\n",
       "      <th>rRNA_depletion</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Monosome_purification</th>\n",
       "      <th>Nuclease</th>\n",
       "      <th>Kit</th>\n",
       "      <th>Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DRR244662</td>\n",
       "      <td>50599092</td>\n",
       "      <td>3845530992</td>\n",
       "      <td>76.0</td>\n",
       "      <td>1273</td>\n",
       "      <td>Ribosome profiling</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AT01_NoTrim_S1_R1_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DRR244663</td>\n",
       "      <td>50768935</td>\n",
       "      <td>3858439060</td>\n",
       "      <td>76.0</td>\n",
       "      <td>1269</td>\n",
       "      <td>Ribosome profiling</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AT02_NoTrim_S2_R1_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DRR255640</td>\n",
       "      <td>53029459</td>\n",
       "      <td>1837645023</td>\n",
       "      <td>34.0</td>\n",
       "      <td>792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribo-seq_WT_rep1_Ribosome footprints</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DRR255641</td>\n",
       "      <td>60864576</td>\n",
       "      <td>2145161510</td>\n",
       "      <td>35.0</td>\n",
       "      <td>874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribo-seq_WT_rep2_Ribosome footprints</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DRR255642</td>\n",
       "      <td>55434616</td>\n",
       "      <td>1923586815</td>\n",
       "      <td>34.0</td>\n",
       "      <td>828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribo-seq_pgr3-1_rep1_Ribosome footprints</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Run     spots       bases  avgLength  size_MB         LibraryName   \n",
       "0  DRR244662  50599092  3845530992       76.0     1273  Ribosome profiling  \\\n",
       "1  DRR244663  50768935  3858439060       76.0     1269  Ribosome profiling   \n",
       "2  DRR255640  53029459  1837645023       34.0      792                 NaN   \n",
       "3  DRR255641  60864576  2145161510       35.0      874                 NaN   \n",
       "4  DRR255642  55434616  1923586815       34.0      828                 NaN   \n",
       "\n",
       "  LibraryStrategy    LibrarySelection   LibrarySource LibraryLayout  ...   \n",
       "0         RNA-Seq              RANDOM  TRANSCRIPTOMIC        SINGLE  ...  \\\n",
       "1         RNA-Seq              RANDOM  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "2         RNA-Seq  size fractionation  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "3         RNA-Seq  size fractionation  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "4         RNA-Seq  size fractionation  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "\n",
       "   Host  UMI Adapter Separation rRNA_depletion Barcode  Monosome_purification   \n",
       "0   NaN  NaN     NaN        NaN            NaN     NaN                    NaN  \\\n",
       "1   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "2   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "3   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "4   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "\n",
       "  Nuclease  Kit                                      Info  \n",
       "0      NaN  NaN                     AT01_NoTrim_S1_R1_001  \n",
       "1      NaN  NaN                     AT02_NoTrim_S2_R1_001  \n",
       "2      NaN  NaN      Ribo-seq_WT_rep1_Ribosome footprints  \n",
       "3      NaN  NaN      Ribo-seq_WT_rep2_Ribosome footprints  \n",
       "4      NaN  NaN  Ribo-seq_pgr3-1_rep1_Ribosome footprints  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2649, 10)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = files[files['study_id'].isin(public_study_ids)]\n",
    "files = files[files['file_type'] == 'riboseq']\n",
    "files.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (303984759.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[10], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    CREATE TABLE `organisms` (\u001b[0m\n\u001b[0m           ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "organisms = pd.read_csv('/home/jack/projects/RiboSeqOrg-DataPortal/data/trips_orgnaims.csv',\n",
    "                        names=['organism_id', 'organism_name', 'transcriptome_list', 'gwips_databasename', 'gwips_clade', 'gwips_organism', 'gwips_database', 'default_transcript', 'private', 'owner'])\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism_id</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>transcriptome_list</th>\n",
       "      <th>gwips_databasename</th>\n",
       "      <th>gwips_clade</th>\n",
       "      <th>gwips_organism</th>\n",
       "      <th>gwips_database</th>\n",
       "      <th>default_transcript</th>\n",
       "      <th>private</th>\n",
       "      <th>owner</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>sgd</td>\n",
       "      <td>yeast</td>\n",
       "      <td>yeast</td>\n",
       "      <td>S.+cerevisiae</td>\n",
       "      <td>sacCer3</td>\n",
       "      <td>YPR122W</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>mycoplasma_hyorhinis</td>\n",
       "      <td>Ensembl_release37</td>\n",
       "      <td>mycoplasma_hyorhinis</td>\n",
       "      <td>mycoplasma_hyorhinis</td>\n",
       "      <td>M.+hyorinis</td>\n",
       "      <td>mh</td>\n",
       "      <td>AFX74150</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>mus_musculus</td>\n",
       "      <td>Gencode_M14</td>\n",
       "      <td>mouse</td>\n",
       "      <td>mammal</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>mm10</td>\n",
       "      <td>ENSMUST00000037796</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>Gencode_v25</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>mammal</td>\n",
       "      <td>Human</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ENST00000558401</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>homo_sapiens_polio</td>\n",
       "      <td>Ensembl_2011</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>mammal</td>\n",
       "      <td>Human</td>\n",
       "      <td>hg38</td>\n",
       "      <td>polio</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   organism_id             organism_name transcriptome_list   \n",
       "0            1  saccharomyces_cerevisiae                sgd  \\\n",
       "1            2      mycoplasma_hyorhinis  Ensembl_release37   \n",
       "2            3              mus_musculus        Gencode_M14   \n",
       "3            4              homo_sapiens        Gencode_v25   \n",
       "4            5        homo_sapiens_polio       Ensembl_2011   \n",
       "\n",
       "     gwips_databasename           gwips_clade gwips_organism gwips_database   \n",
       "0                 yeast                 yeast  S.+cerevisiae        sacCer3  \\\n",
       "1  mycoplasma_hyorhinis  mycoplasma_hyorhinis    M.+hyorinis             mh   \n",
       "2                 mouse                mammal          Mouse           mm10   \n",
       "3          homo_sapiens                mammal          Human           hg38   \n",
       "4          homo_sapiens                mammal          Human           hg38   \n",
       "\n",
       "   default_transcript  private  owner  \n",
       "0             YPR122W        0      1  \n",
       "1            AFX74150        1      1  \n",
       "2  ENSMUST00000037796        0      1  \n",
       "3     ENST00000558401        0      1  \n",
       "4               polio        1      1  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organisms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/jack/projects/RiboSeqOrg-DataPortal/data/trips_model.tsv', 'w') as f:\n",
    "    for row in files.iterrows():\n",
    "        match = filtered[filtered['Run'] == row[1]['file_name'].split('.')[0]]\n",
    "        if len(match['Run'].values) == 0:\n",
    "            run = 'NA'\n",
    "            bioproject = 'NA'\n",
    "        else:\n",
    "            run = match['Run'].values[0]\n",
    "            bioproject = match['BioProject'].values[0]\n",
    "        organism = organisms[organisms['organism_id'] == row[1]['organism_id']]\n",
    "        if organism.shape[0] == 0:\n",
    "            continue\n",
    "        study = studies[studies['study_id'] == row[1]['study_id']]\n",
    "        f.write(f\"{bioproject}\\t{run}\\t{row[1]['file_id']}\\t{row[1]['file_name'].split('.')[0]}\\t{study['study_name'].values[0]}\\t{study['srp_nos'].values[0]}\\t{study['gse_nos'].values[0]}\\t{study['paper_pmid'].values[0]}\\t{organism['organism_name'].values[0]}\\t{organism['transcriptome_list'].values[0]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism_id</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>transcriptome_list</th>\n",
       "      <th>gwips_databasename</th>\n",
       "      <th>gwips_clade</th>\n",
       "      <th>gwips_organism</th>\n",
       "      <th>gwips_database</th>\n",
       "      <th>default_transcript</th>\n",
       "      <th>private</th>\n",
       "      <th>owner</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [organism_id, organism_name, transcriptome_list, gwips_databasename, gwips_clade, gwips_organism, gwips_database, default_transcript, private, owner]\n",
       "Index: []"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organisms[organisms['organism_id'] == 15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_106108/2695790450.py:2: DtypeWarning: Columns (37,39,40,41,63,72,73,78,81,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"/home/jack/projects/RiboSeqOrg-DataPortal/data/filtered_riboseq_done_260623.csv\")\n"
     ]
    }
   ],
   "source": [
    "updated_df = pd.read_csv(\"/home/jack/projects/RiboSeqOrg-DataPortal/data/metadata_improved_with_manual.csv\")\n",
    "df = pd.read_csv(\"/home/jack/projects/RiboSeqOrg-DataPortal/data/filtered_riboseq_done_260623.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Run', 'spots', 'bases', 'avgLength', 'size_MB', 'LibraryName',\n",
       "       'LibraryStrategy', 'LibrarySelection', 'LibrarySource', 'LibraryLayout',\n",
       "       'InsertSize', 'InsertDev', 'Platform', 'Model', 'SRAStudy',\n",
       "       'BioProject', 'Study_Pubmed_id', 'Sample', 'BioSample', 'SampleType',\n",
       "       'TaxID', 'ScientificName', 'SampleName', 'CenterName', 'Submission',\n",
       "       'MONTH', 'YEAR', 'AUTHOR', 'sample_source', 'sample_title',\n",
       "       'ENA_first_public', 'ENA_last_update', 'INSDC_center_alias',\n",
       "       'INSDC_center_name', 'INSDC_first_public', 'INSDC_last_update',\n",
       "       'INSDC_status', 'ENA_checklist', 'GEO_Accession', 'Experiment_Date',\n",
       "       'date_sequenced', 'submission_date', 'date', 'Experiment', 'CELL_LINE',\n",
       "       'TISSUE', 'INHIBITOR', 'TIMEPOINT', 'FRACTION', 'REPLICATE',\n",
       "       'CONDITION', 'LIBRARYTYPE', 'STAGE', 'GENE', 'Sex', 'Strain', 'Age',\n",
       "       'Infected', 'Disease', 'Genotype', 'Feeding', 'Temperature', 'SiRNA',\n",
       "       'SgRNA', 'ShRNA', 'Plasmid', 'Growth_Condition', 'Stress', 'Cancer',\n",
       "       'microRNA', 'Individual', 'Antibody', 'Ethnicity', 'Dose',\n",
       "       'Stimulation', 'Host', 'UMI', 'Adapter', 'Separation', 'rRNA_depletion',\n",
       "       'Barcode', 'Monosome_purification', 'Nuclease', 'Kit', 'Info'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13800, 85)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>spots</th>\n",
       "      <th>bases</th>\n",
       "      <th>avgLength</th>\n",
       "      <th>size_MB</th>\n",
       "      <th>LibraryName</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>...</th>\n",
       "      <th>Host</th>\n",
       "      <th>UMI</th>\n",
       "      <th>Adapter</th>\n",
       "      <th>Separation</th>\n",
       "      <th>rRNA_depletion</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Monosome_purification</th>\n",
       "      <th>Nuclease</th>\n",
       "      <th>Kit</th>\n",
       "      <th>Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DRR244662</td>\n",
       "      <td>50599092</td>\n",
       "      <td>3845530992</td>\n",
       "      <td>76.0</td>\n",
       "      <td>1273</td>\n",
       "      <td>Ribosome profiling</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AT01_NoTrim_S1_R1_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DRR244663</td>\n",
       "      <td>50768935</td>\n",
       "      <td>3858439060</td>\n",
       "      <td>76.0</td>\n",
       "      <td>1269</td>\n",
       "      <td>Ribosome profiling</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AT02_NoTrim_S2_R1_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DRR255640</td>\n",
       "      <td>53029459</td>\n",
       "      <td>1837645023</td>\n",
       "      <td>34.0</td>\n",
       "      <td>792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribo-seq_WT_rep1_Ribosome footprints</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DRR255641</td>\n",
       "      <td>60864576</td>\n",
       "      <td>2145161510</td>\n",
       "      <td>35.0</td>\n",
       "      <td>874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribo-seq_WT_rep2_Ribosome footprints</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DRR255642</td>\n",
       "      <td>55434616</td>\n",
       "      <td>1923586815</td>\n",
       "      <td>34.0</td>\n",
       "      <td>828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribo-seq_pgr3-1_rep1_Ribosome footprints</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Run     spots       bases  avgLength  size_MB         LibraryName   \n",
       "0  DRR244662  50599092  3845530992       76.0     1273  Ribosome profiling  \\\n",
       "1  DRR244663  50768935  3858439060       76.0     1269  Ribosome profiling   \n",
       "2  DRR255640  53029459  1837645023       34.0      792                 NaN   \n",
       "3  DRR255641  60864576  2145161510       35.0      874                 NaN   \n",
       "4  DRR255642  55434616  1923586815       34.0      828                 NaN   \n",
       "\n",
       "  LibraryStrategy    LibrarySelection   LibrarySource LibraryLayout  ...   \n",
       "0         RNA-Seq              RANDOM  TRANSCRIPTOMIC        SINGLE  ...  \\\n",
       "1         RNA-Seq              RANDOM  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "2         RNA-Seq  size fractionation  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "3         RNA-Seq  size fractionation  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "4         RNA-Seq  size fractionation  TRANSCRIPTOMIC        SINGLE  ...   \n",
       "\n",
       "   Host  UMI Adapter Separation rRNA_depletion Barcode  Monosome_purification   \n",
       "0   NaN  NaN     NaN        NaN            NaN     NaN                    NaN  \\\n",
       "1   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "2   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "3   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "4   NaN  NaN     NaN        NaN            NaN     NaN                    NaN   \n",
       "\n",
       "  Nuclease  Kit                                      Info  \n",
       "0      NaN  NaN                     AT01_NoTrim_S1_R1_001  \n",
       "1      NaN  NaN                     AT02_NoTrim_S2_R1_001  \n",
       "2      NaN  NaN      Ribo-seq_WT_rep1_Ribosome footprints  \n",
       "3      NaN  NaN      Ribo-seq_WT_rep2_Ribosome footprints  \n",
       "4      NaN  NaN  Ribo-seq_pgr3-1_rep1_Ribosome footprints  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_rows = []\n",
    "for _, row in df.iterrows():\n",
    "    updated_df_row = updated_df[updated_df['Run'] == row['Run']]\n",
    "    if updated_df_row.shape[0] == 0:\n",
    "        clean_rows.append(row)\n",
    "        continue\n",
    "    if str(updated_df_row['TISSUE_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['TISSUE'] = updated_df_row['TISSUE_st'].values[0]\n",
    "    if str(updated_df_row['CELL_LINE_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['CELL_LINE'] = updated_df_row['CELL_LINE_st'].values[0]\n",
    "    if str(updated_df_row['INHIBITOR_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['INHIBITOR'] = updated_df_row['INHIBITOR_st'].values[0]\n",
    "    if str(updated_df_row['CONDITION_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['CONDITION'] = updated_df_row['CONDITION_st'].values[0]\n",
    "    if str(updated_df_row['REPLICATE_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['REPLICATE'] = updated_df_row['REPLICATE_st'].values[0]\n",
    "    if str(updated_df_row['LIBRARYTYPE_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['LIBRARYTYPE'] = updated_df_row['LIBRARYTYPE_st'].values[0]\n",
    "    if str(updated_df_row['FRACTION_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['FRACTION'] = updated_df_row['FRACTION_st'].values[0]\n",
    "    if str(updated_df_row['TIMEPOINT_st'].values[0]) not in ['nan', \"NA\", '']:\n",
    "        row['TIMEPOINT'] = updated_df_row['TIMEPOINT_st'].values[0]\n",
    "    # row.drop(['TISSUE_st', 'CELL_LINE_st', 'INHIBITOR_st', 'CONDITION_st', 'REPLICATE_st', 'LIBRARYTYPE_st', 'FRACTION_st', 'TIMEPOINT_st'], inplace=True)\n",
    "    clean_rows.append(row)\n",
    "\n",
    "clean_df = pd.DataFrame(clean_rows)\n",
    "clean_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LIBRARYTYPE_st\n",
       "RFP        10348\n",
       "RNA          603\n",
       "RMS          195\n",
       "40S           77\n",
       "LSU           60\n",
       "RiboTag       42\n",
       "RIP           16\n",
       "SSU           11\n",
       "tRNA           5\n",
       "Ribotag        4\n",
       "CLIP           4\n",
       "CRAC           4\n",
       "CAGE           2\n",
       "TRAP           1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "updated_df['LIBRARYTYPE_st'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Run': 0.0,\n",
       " 'spots': 0.0,\n",
       " 'bases': 0.0,\n",
       " 'avgLength': 0.2101449275362319,\n",
       " 'size_MB': 0.0,\n",
       " 'LibraryName': 73.23913043478261,\n",
       " 'LibraryStrategy': 0.0,\n",
       " 'LibrarySelection': 0.0,\n",
       " 'LibrarySource': 0.0,\n",
       " 'LibraryLayout': 0.0,\n",
       " 'InsertSize': 0.0,\n",
       " 'InsertDev': 0.0,\n",
       " 'Platform': 0.0,\n",
       " 'Model': 0.0,\n",
       " 'SRAStudy': 0.0,\n",
       " 'BioProject': 0.0,\n",
       " 'Study_Pubmed_id': 0.0,\n",
       " 'Sample': 0.0,\n",
       " 'BioSample': 0.0,\n",
       " 'SampleType': 0.0,\n",
       " 'TaxID': 0.0,\n",
       " 'ScientificName': 0.2246376811594203,\n",
       " 'SampleName': 0.0,\n",
       " 'CenterName': 0.5362318840579711,\n",
       " 'Submission': 0.0,\n",
       " 'MONTH': 0.0,\n",
       " 'YEAR': 0.0,\n",
       " 'AUTHOR': 0.0,\n",
       " 'sample_source': 17.07246376811594,\n",
       " 'sample_title': 5.282608695652174,\n",
       " 'ENA_first_public': 94.03623188405797,\n",
       " 'ENA_last_update': 94.03623188405797,\n",
       " 'INSDC_center_alias': 93.4927536231884,\n",
       " 'INSDC_center_name': 93.4927536231884,\n",
       " 'INSDC_first_public': 93.4927536231884,\n",
       " 'INSDC_last_update': 93.4927536231884,\n",
       " 'INSDC_status': 93.4927536231884,\n",
       " 'ENA_checklist': 99.94202898550725,\n",
       " 'GEO_Accession': 28.8695652173913,\n",
       " 'Experiment_Date': 99.97101449275362,\n",
       " 'date_sequenced': 99.90579710144928,\n",
       " 'submission_date': 99.95652173913044,\n",
       " 'date': 0.0,\n",
       " 'Experiment': 0.0,\n",
       " 'CELL_LINE': 64.71014492753623,\n",
       " 'TISSUE': 34.7536231884058,\n",
       " 'INHIBITOR': 72.59420289855072,\n",
       " 'TIMEPOINT': 85.31159420289856,\n",
       " 'FRACTION': 93.42028985507247,\n",
       " 'REPLICATE': 88.81884057971014,\n",
       " 'CONDITION': 94.8768115942029,\n",
       " 'LIBRARYTYPE': 68.67391304347827,\n",
       " 'STAGE': 82.53623188405797,\n",
       " 'GENE': 95.3768115942029,\n",
       " 'Sex': 91.77536231884058,\n",
       " 'Strain': 51.23188405797101,\n",
       " 'Age': 82.33333333333334,\n",
       " 'Infected': 96.83333333333334,\n",
       " 'Disease': 99.70289855072464,\n",
       " 'Genotype': 61.28985507246377,\n",
       " 'Feeding': 93.31159420289855,\n",
       " 'Temperature': 98.2463768115942,\n",
       " 'SiRNA': 99.63768115942028,\n",
       " 'SgRNA': 99.84057971014492,\n",
       " 'ShRNA': 99.42028985507247,\n",
       " 'Plasmid': 97.89855072463767,\n",
       " 'Growth_Condition': 87.66666666666667,\n",
       " 'Stress': 96.9927536231884,\n",
       " 'Cancer': 98.3623188405797,\n",
       " 'microRNA': 99.59420289855072,\n",
       " 'Individual': 97.37681159420289,\n",
       " 'Antibody': 99.34782608695653,\n",
       " 'Ethnicity': 99.84057971014492,\n",
       " 'Dose': 98.4927536231884,\n",
       " 'Stimulation': 99.23188405797102,\n",
       " 'Host': 99.12318840579711,\n",
       " 'UMI': 99.3840579710145,\n",
       " 'Adapter': 99.02898550724638,\n",
       " 'Separation': 96.07971014492753,\n",
       " 'rRNA_depletion': 97.83333333333334,\n",
       " 'Barcode': 98.52898550724638,\n",
       " 'Monosome_purification': 99.97101449275362,\n",
       " 'Nuclease': 99.46376811594203,\n",
       " 'Kit': 99.20289855072464,\n",
       " 'Info': 4.840579710144928}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nan_percentages = clean_df.isna().mean() * 100\n",
    "dict(nan_percentages)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_df.loc[clean_df['Study_Pubmed_id'] == 1, 'Study_Pubmed_id'] = np.nan\n",
    "clean_df.loc[clean_df['AUTHOR'] == 'Makar', 'AUTHOR'] = np.nan\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Study_Pubmed_id\n",
       "26554015.0    884\n",
       "24705786.0    492\n",
       "34316693.0    200\n",
       "29198561.0    165\n",
       "35626002.0    158\n",
       "             ... \n",
       "32851185.0      1\n",
       "27938667.0      1\n",
       "29107537.0      1\n",
       "31259691.0      1\n",
       "26876183.0      1\n",
       "Name: count, Length: 571, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df['Study_Pubmed_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AUTHOR\n",
       "Atger       884\n",
       "Bazzini     526\n",
       "Marcel      200\n",
       "Chou        165\n",
       "Mohan       158\n",
       "           ... \n",
       "Hickey        2\n",
       "Weinberg      1\n",
       "Jin           1\n",
       "Hanson        1\n",
       "Gupta         1\n",
       "Name: count, Length: 445, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df['AUTHOR'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Run': 0.0,\n",
       " 'spots': 0.0,\n",
       " 'bases': 0.0,\n",
       " 'avgLength': 0.2101449275362319,\n",
       " 'size_MB': 0.0,\n",
       " 'LibraryName': 73.23913043478261,\n",
       " 'LibraryStrategy': 0.0,\n",
       " 'LibrarySelection': 0.0,\n",
       " 'LibrarySource': 0.0,\n",
       " 'LibraryLayout': 0.0,\n",
       " 'InsertSize': 0.0,\n",
       " 'InsertDev': 0.0,\n",
       " 'Platform': 0.0,\n",
       " 'Model': 0.0,\n",
       " 'SRAStudy': 0.0,\n",
       " 'BioProject': 0.0,\n",
       " 'Study_Pubmed_id': 0.0,\n",
       " 'Sample': 0.0,\n",
       " 'BioSample': 0.0,\n",
       " 'SampleType': 0.0,\n",
       " 'TaxID': 0.0,\n",
       " 'ScientificName': 0.2246376811594203,\n",
       " 'SampleName': 0.0,\n",
       " 'CenterName': 0.5362318840579711,\n",
       " 'Submission': 0.0,\n",
       " 'MONTH': 0.0,\n",
       " 'YEAR': 0.0,\n",
       " 'AUTHOR': 0.0,\n",
       " 'sample_source': 17.07246376811594,\n",
       " 'sample_title': 5.282608695652174,\n",
       " 'ENA_first_public': 94.03623188405797,\n",
       " 'ENA_last_update': 94.03623188405797,\n",
       " 'INSDC_center_alias': 93.4927536231884,\n",
       " 'INSDC_center_name': 93.4927536231884,\n",
       " 'INSDC_first_public': 93.4927536231884,\n",
       " 'INSDC_last_update': 93.4927536231884,\n",
       " 'INSDC_status': 93.4927536231884,\n",
       " 'ENA_checklist': 99.94202898550725,\n",
       " 'GEO_Accession': 28.8695652173913,\n",
       " 'Experiment_Date': 99.97101449275362,\n",
       " 'date_sequenced': 99.90579710144928,\n",
       " 'submission_date': 99.95652173913044,\n",
       " 'date': 0.0,\n",
       " 'Experiment': 0.0,\n",
       " 'CELL_LINE': 64.71014492753623,\n",
       " 'TISSUE': 65.05797101449275,\n",
       " 'INHIBITOR': 72.59420289855072,\n",
       " 'TIMEPOINT': 85.31159420289856,\n",
       " 'FRACTION': 93.42028985507247,\n",
       " 'REPLICATE': 88.81884057971014,\n",
       " 'CONDITION': 94.8768115942029,\n",
       " 'LIBRARYTYPE': 68.67391304347827,\n",
       " 'STAGE': 82.53623188405797,\n",
       " 'GENE': 95.3768115942029,\n",
       " 'Sex': 91.77536231884058,\n",
       " 'Strain': 51.23188405797101,\n",
       " 'Age': 82.33333333333334,\n",
       " 'Infected': 96.83333333333334,\n",
       " 'Disease': 99.70289855072464,\n",
       " 'Genotype': 61.28985507246377,\n",
       " 'Feeding': 93.31159420289855,\n",
       " 'Temperature': 98.2463768115942,\n",
       " 'SiRNA': 99.63768115942028,\n",
       " 'SgRNA': 99.84057971014492,\n",
       " 'ShRNA': 99.42028985507247,\n",
       " 'Plasmid': 97.89855072463767,\n",
       " 'Growth_Condition': 87.66666666666667,\n",
       " 'Stress': 96.9927536231884,\n",
       " 'Cancer': 98.3623188405797,\n",
       " 'microRNA': 99.59420289855072,\n",
       " 'Individual': 97.37681159420289,\n",
       " 'Antibody': 99.34782608695653,\n",
       " 'Ethnicity': 99.84057971014492,\n",
       " 'Dose': 98.4927536231884,\n",
       " 'Stimulation': 99.23188405797102,\n",
       " 'Host': 99.12318840579711,\n",
       " 'UMI': 99.3840579710145,\n",
       " 'Adapter': 99.02898550724638,\n",
       " 'Separation': 96.07971014492753,\n",
       " 'rRNA_depletion': 97.83333333333334,\n",
       " 'Barcode': 98.52898550724638,\n",
       " 'Monosome_purification': 99.97101449275362,\n",
       " 'Nuclease': 99.46376811594203,\n",
       " 'Kit': 99.20289855072464,\n",
       " 'Info': 4.840579710144928}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nan_percentages = df.isna().mean() * 100\n",
    "dict(nan_percentages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CELL_LINE\n",
       "HEK293T                                    302\n",
       "293T4A_Human embryonic kidney cell line    158\n",
       "HeLa                                       139\n",
       "A549                                       132\n",
       "HEK293                                      86\n",
       "                                          ... \n",
       "12-cell                                      1\n",
       "5-cell                                       1\n",
       "6-cell                                       1\n",
       "GM18856_Lymphoblastoid Cell Line             1\n",
       "GM19193_EBV-transformed lymphoblastoid       1\n",
       "Name: count, Length: 561, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df['CELL_LINE'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_df.to_csv(\"/home/jack/projects/RiboSeqOrg-DataPortal/data/filtered_standardised_07_07_23.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13800, 85)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_df.shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RiboSeqOrg-DataPortal",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
